### PR TITLE
Saving

### PR DESCRIPTION
**Pull request by graberyanis**
- added file manager for opening and reading from/loading to files
- main: 
- uses file manager to open file, saves networks and generation counter into dictionary after ending the simulation and loads dictionary to json file.
- layer:
- layers can be given inputs for weights and biases, for using save data, if none are given, creates random weights/biases
- network:
- networks can be given weights/biases as inputs which will be used to create the single layers
- population:
- population can be given a list with weights and biases of another generation which will be used to create a new population
- simulation:
- uses save dictionary for creating the population and setting the generation counter, simulation will now only stop after the current generation has ended (to avoid distortion in data) (will later also change, so population will only be reset if the simulation is running)

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -6,10 +6,24 @@
 #from evolution.reproduction import reproduce
 #from evolution.mutation import mutate
 from simulation import Simulation
+from utils.file_manager import FileManager
 
 def main():
-    simulation = Simulation()
+    FM = FileManager()
```
Comment (webermael) [inline src/main.py]: Would you mind making this lowercase? Just to keep everything _uniform_ ...

**Review by webermael — APPROVED** [review]
Verry clean code (and everything still works 🙃)

**Comment by graberyanis** [comment]
The config file has been made redundant by the almighty input_dict, who came from the far land named: template.json
Also fire is now really burning with a cool new color changing effect!
Fire won't set fire on fire anymore
Bigger better and blue organisms for more enjoyement (also now equipped with radars to see where they are :D)

File: src/evolution/mutation.py [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import random
-def mutation(config, value):
-    if random.random() <= config.mutation_rate:
-        return value + random.uniform(config.mutation_max_difference, -config.mutation_max_difference)
+def mutation(inputs, value):
+    if random.random() <= inputs["chance"]:
```
Comment (webermael) [inline src/evolution/mutation.py]: Shouldn't it be called "probability" instead of "chance"?

File: src/main.py [hunk 3]
```diff
@@ -1,15 +1,21 @@
-#import pygame
-#from config import Config
-#from world.world import World
-#from organisms.population import Population
-#from utils.visualisation import (render_world, screen_init, screen_update)
-#from evolution.reproduction import reproduce
-#from evolution.mutation import mutate
```
Comment (webermael) [inline src/main.py:7]: at last

File: src/organisms/organism.py [hunk 4]
```diff
@@ -29,37 +29,37 @@ def __init__(self, config, x_pos, y_pos, nn = False):
...
+        if 0 <= self.position.x + direction[0] < inputs["world"]["size"]["row_length"] and 0 <= self.position.y + direction[1] < inputs["world"]["size"]["column_length"]:
             if not world[self.position.y + direction[1]][self.position.x + direction[0]].has_organism:
                 self.position.x += direction[0]
                 self.position.y += direction[1]
                 if not direction == (0, 0):
-                    self.score += config.move_bonus
+                    self.score += inputs["population"]["interactions"]["move_bonus"]
         else:
-            self.lifetime -= config.border_damage
+            self.lifetime -= inputs["population"]["interactions"]["border_damage"]
```
Comment (webermael) [inline src/organisms/organism.py:44]: Just to throw it in there: do we still want to take damage at the border?
I think we once talked about "wrapping around", so the organisms can move "freely"...
-> What did we decide there?

File: src/organisms/sensors.py [hunk 5]
```diff
@@ -4,15 +4,15 @@ def __init__(self, size):
         self.translation = {"danger" : 0, "empty" : 0.4, "food" : 1}
         self.org_translation = {"danger" : 0, "empty" : 0.6, "food" : 0.8}
```
Comment (webermael) [inline src/organisms/sensors.py:5]: What are we doing about that now? I don't think we really came to a conclusion on this.

**Review by webermael — APPROVED** [review]
Danger looks _fire_

File: src/evolution/mutation.py (hunk 2 above)
Comment (graberyanis) [reply src/evolution/mutation.py]: good point

File: src/organisms/organism.py (hunk 4 above)
Comment (graberyanis) [reply src/organisms/organism.py:44]: I would leave it for now, maybe change it after having the branch and the interface together...

File: src/organisms/sensors.py (hunk 5 above)
Comment (graberyanis) [reply src/organisms/sensors.py:5]: I feel like it worked out, the biggest issue i feel like is them not comprehending the edge of the world

**Pull request merged**